### PR TITLE
Fix file move/copy when storage space is not enough

### PR DESCRIPTION
**Pull request by PVince81**
- properly detect streamCopy errors
- when a streamCopy failed, delete the partially written target file

This applies for both "rename" and "copy"

@icewind1991 @DeepDiver1975 @schiesbn

### CODE REVIEW

**Comment by PVince81** [comment]
Fixes https://github.com/owncloud/core/issues/14284

Backport PR for stable8 here: https://github.com/owncloud/core/pull/14504

File: lib/private/connector/sabre/objecttree.php [hunk 1]
```diff
@@ -236,7 +236,9 @@ public function copy($source, $destination) {
 
 		try {
 			if ($this->fileView->is_file($source)) {
```
Comment (icewind1991) [inline lib/private/connector/sabre/objecttree.php]: We can remove this check since $view->copy works on folders

Comment (PVince81) [inline lib/private/connector/sabre/objecttree.php]: Will do as part of https://github.com/owncloud/core/issues/14501

File: lib/private/files/view.php [hunk 2]
```diff
@@ -656,6 +662,11 @@ public function rename($path1, $path2) {
 
 						if ($result !== false) {
 							$result &= $storage1->unlink($internalPath1);
+						} else {
+							// delete partially written target file
+							$storage2->unlink($internalPath2);
+							// delete cache entry that was created by fopen
+							$storage2->getCache()->remove($internalPath2);
```
Comment (icewind1991) [inline lib/private/files/view.php]: No need for this I think, cache hasn't been updated at this point

File: lib/private/connector/sabre/objecttree.php (hunk 1 above)
Comment (PVince81) [inline lib/private/connector/sabre/objecttree.php]: Here we go https://github.com/owncloud/core/pull/14505

File: lib/private/files/view.php (hunk 2 above)
Comment (PVince81) [inline lib/private/files/view.php]: It has unfortunately. The fopen() operation sometimes updates the cache. (observed locally)

Comment (PVince81) [inline lib/private/files/view.php]: Try commenting it out and observe the failing unit test

**Comment by PVince81** [comment]
@owncloud-bot retest this please

**Comment by PVince81** [comment]
@owncloud-bot retest this please

**Comment by PVince81** [comment]
Can this be merged ? @icewind1991 @DeepDiver1975 @nickvergessen @MorrisJobke @LukasReschke

File: lib/private/connector/sabre/objecttree.php [hunk 3]
```diff
@@ -236,7 +236,9 @@ public function copy($source, $destination) {
...
-				$this->fileView->copy($source, $destination);
+				if (!$this->fileView->copy($source, $destination)) {
+					throw new \Sabre\DAV\Exception\Forbidden('');
```
Comment (LukasReschke) [inline lib/private/connector/sabre/objecttree.php]: PHPDoc needs an update as well: 

screen shot 2015-02-26 at 12 27 02

File: lib/private/connector/sabre/objecttree.php [hunk 4]
```diff
@@ -244,7 +246,9 @@ public function copy($source, $destination) {
 					while (($subNode = readdir($dh)) !== false) {
 
 						if ($subNode == '.' || $subNode == '..') continue;
-						$this->copy($source . '/' . $subNode, $destination . '/' . $subNode);
+						if (!$this->copy($source . '/' . $subNode, $destination . '/' . $subNode)) {
```
Comment (LukasReschke) [inline lib/private/connector/sabre/objecttree.php]: Copy is a void method? – I don't get this. Did you mean to use a try-catch block instead?

File: lib/private/files/view.php [hunk 5]
```diff
@@ -574,10 +574,14 @@ public function deleteAll($directory) {
...
-	 * @param string $path2
+	 * Rename/move a file or folder from the source path to target path.
+	 *
+	 * @param string $path1 source path
+	 * @param string $path2 target path
+	 *
 	 * @return bool|mixed
-	 */
+	 * 
+	 **/
```
Comment (LukasReschke) [inline lib/private/files/view.php]: Ending block has only 1 `*`

Comment (LukasReschke) [inline lib/private/files/view.php]: Also newline looks off

File: lib/private/connector/sabre/objecttree.php (hunk 4 above)
Comment (PVince81) [inline lib/private/connector/sabre/objecttree.php]: Damn, I thought this was "$this->fileView->copy()".

Comment (PVince81) [inline lib/private/connector/sabre/objecttree.php]: I'll revert this. It will be gone anyway as part of https://github.com/owncloud/core/pull/14505

**Comment by PVince81** [comment]
@LukasReschke fixed. (I discarded the unneeded changes in objectree)

**Comment by MorrisJobke** [comment]
:+1:

**Comment by LukasReschke** [comment]
Thanks! :+1:

**Pull request merged**